main/ZgatewayBT.ino

```diff
@@ -374,146 +381,34 @@ class MyAdvertisedDeviceCallbacks : public BLEAdvertisedDeviceCallbacks
-              }
-              pos = -1;
-              pos = strpos(service_data, "4030dd");
-              if (pos != -1)
-              {
-                Log.trace(F("Mi Lamp data reading" CR));
-                //example "servicedata":4030DD031D0300010100
-
-                if (!isDiscovered(device))
-                  MiLampDiscovery(mac);
-
-                process_milamp(service_data, mac);
-              }
-            }
-            if (strstr(BLEdata["servicedatauuid"].as<char *>(), "181d") != NULL)
-            { // Mi Scale V1
-              Log.trace(F("Mi Scale V1 data reading" CR));
-              //example "servicedata":"a2ac2be307060207122b" /"a28039e3070602070e28"
-
-              if (!isDiscovered(device))
-                MiScaleDiscovery(mac);
-
-              process_scale_v1(service_data, mac);
-            }
-            if (strstr(BLEdata["servicedatauuid"].as<char *>(), "181b") != NULL)
-            { // Mi Scale V2
-              Log.trace(F("Mi Scale V2 data reading" CR));
-              //example "servicedata":02c4e1070b1e13050c00002607 / 02a6e20705150a251df401443e /02a6e20705180c0d04d701943e
-
-              if (!isDiscovered(device))
-                MiScaleDiscovery(mac);
-
-              process_scale_v2(service_data, mac);
-            }
-            if (strstr(BLEdata["servicedatauuid"].as<char *>(), "fee0") != NULL)
-            { // Mi Band //0000fee0-0000-1000-8000-00805f9b34fb // ESP32 only
-              Log.trace(F("Mi Band data reading" CR));
-              //example "servicedata":a21e0000
-
-              if (!isDiscovered(device))
-                MiBandDiscovery(mac);
-
-              process_miband(service_data, mac);
-            }
-            if (strstr(BLEdata["servicedata"].as<char *>(), "08094c") != NULL)
-            { // Clear grass with air pressure//08094c0140342d580104d8000c020702612702015a
-              Log.trace(F("Clear grass data with air pressure reading" CR));
-              //example "servicedata":08094c0140342d580104 c400 2402 0702 5d27 02015a
-
-              if (!isDiscovered(device))
-                CLEARGRASSTRHKPADiscovery(mac);
-
-              process_cleargrass_air(service_data, mac);
-            }
-            if (strstr(BLEdata["servicedata"].as<char *>(), "080774") != NULL)
-            { // Clear grass standard method 2/0807743e10342d580104c3002c0202012a
-              Log.trace(F("Clear grass data reading method 2" CR));
-              //example "servicedata":0807743e10342d580104 c300 2c02 02012a
-              // no discovery as it is already available with method 1
-              process_cleargrass(service_data, mac);
+          { // process only the devices close enough
+            JsonObject &BLEdataOut = process_bledata(BLEdata);
+            #ifdef ZmqttDiscovery
+              launch_discovery(BLEdataOut, mac);
+            #endif
+            #if !pubBLEServiceUUID
+            if(BLEdataOut.containsKey("servicedatauuid"))
```

Comment:
Line 396-400
Line 403-407
Line 672-676

At first this can be extracted as a function or macro with the signature (psydocode): 

```
void RemoveJsonPropertyIf(JsonObject &obj, char* key, bool condition)
{
  if(condition){
     Log.trace(F("Removing %s" CR), key);
     obj.Remove(key);
  }
}
```

but a second, all of this postprocessing can be merged and shared between ESP32 and HM10.

I mean Line 389-416 are mostly the same as Line 665-686.

```
void PublishDeviceData(JsonObject &BLEdata)
{
          if (abs((int)BLEdata["rssi"] | 0) < abs(Minrssi))
          { // process only the devices close enough
            JsonObject &BLEdataOut = process_bledata(BLEdata);
            #ifdef ZmqttDiscovery
              launch_discovery(BLEdataOut, mac);
            #endif
            #if !pubBLEServiceUUID
              RemoveJsonPropertyIf(BLEdataOut, "servicedatauuid", BLEdataOut.containsKey("servicedatauuid"));
            #endif
            #if !pubKnownBLEServiceData
              RemoveJsonPropertyIf(BLEdataOut, "servicedata", BLEdataOut.containsKey("model") && BLEdataOut.containsKey("servicedata"));
            #endif
            String mactopic(mac);
            mactopic = subjectBTtoMQTT + String("/") + mactopic;
            pub((char *)mactopic.c_str(), BLEdataOut);
          }
          else
          {
            Log.trace(F("Low rssi, device filtered" CR));
          }
        }
}
```

or maybe this would be placed into `process_bledata`

Reply:
Thanks for this valuable feedback. I'm implementing the changes.